src/jit/codegenxarch.cpp

```diff
@@ -4247,241 +4247,133 @@ instruction CodeGen::genGetInsForOper(genTreeOps oper, var_types type)
+        // sit in ECX. In case this didn't happen, LSRA expects the code generator to move it since it's a single
+        // register destination requirement.
+        regNumber shiftReg = shiftBy->gtRegNum;
+        if (shiftReg != REG_RCX)
         {
-            operandReg = operand->gtRegNum;
+            // Issue the mov to RCX:
+            inst_RV_RV(INS_mov, REG_RCX, shiftReg, shiftBy->TypeGet());
+            shiftReg = REG_RCX;
         }
-        else if (actualOperand->OperGet() == GT_LEA)
-        {
-            operandReg = actualOperand->gtOp.gtOp1->gtRegNum;
-            GenTreeAddrMode* addrMode = actualOperand->AsAddrMode();
-            offset = addrMode->gtOffset;
-            if(addrMode->Index() != nullptr)
-            {
-                indexReg = addrMode->Index()->gtRegNum;
 
-                // GT_LEA with an indexReg is not supported for shift by immediate
-                assert(!shiftBy->isContainedIntOrIImmed());
-            }
-        }
-        else if (actualOperand->IsCnsIntOrI())
+        // The operand to be shifted must not be in ECX
+        noway_assert(operandReg != REG_RCX);
+
+        if (parent->gtRegNum != operandReg)
         {
-            GenTreeIntConCommon* intCon = actualOperand->AsIntConCommon();
-            if (actualOperand->isContained())
-            {
-                // Contained absolute address should fit within addr base
-                assert(intCon->FitsInAddrBase(compiler));
+            inst_RV_RV(INS_mov, parent->gtRegNum, operandReg, targetType);
+        }
+        inst_RV_CL(ins, parent->gtRegNum, targetType);
+    }
 
-                // Don't expect to see GT_COPY or GT_RELOAD
-                assert(operand == actualOperand);
+    genProduceReg(parent);
+}
 
-                isCnsIntOrIAndFitsWithinAddrBase = true;
-                disp = intCon->IconValue(); 
 
-                if (intCon->AddrNeedsReloc(compiler))
-                {
-                    attr = EA_SET_FLG(attr, EA_DSP_RELOC_FLG);
-                }
-            }
-            else
-            {
-                operandReg = operand->gtRegNum;
-            }
-        }
-        else 
-        {
-            // The only other supported operands for RMW are GT_CLS_VAR_ADDR and GT_LCL_VAR_ADDR
-            assert(actualOperand->OperGet() == GT_CLS_VAR_ADDR || actualOperand->OperGet() == GT_LCL_VAR_ADDR);
+//------------------------------------------------------------------------
+// genCodeForShiftRMW: Generates the code sequence for a GT_STOREIND GenTree node that
+// represents a RMW bit shift or rotate operation (<<, >>, >>>, rol, ror), for example:
+//      GT_STOREIND( AddressTree, GT_SHL( Ind ( AddressTree ), Operand ) )
+//
+// Arguments:
+//    storeIndNode: the GT_STOREIND node.
+//
+void CodeGen::genCodeForShiftRMW(GenTreePtr storeIndNode)
+{
+    GenTreeStoreInd* storeInd = storeIndNode->AsStoreInd();
```

Comment:
Since the caller of this routine is has a GenTreeStoreInd*, we can make type of param of this routine to be GenTreeStoreInd* and avoid the need to cast it using AsStoreInd()